src/ros_filter.cpp

```diff
@@ -2971,12 +2973,32 @@ namespace RobotLocalization
 
     // Walk back through the queue until we reach a filter state whose time stamp is less than or equal to the
     // requested time. Since every saved state after that time will be overwritten/corrected, we can pop from
-    // the queue.
+    // the queue. If the history is insufficiently short, we just take the oldest state we have.
+    FilterStatePtr lastHistoryState;
     while (!filterStateHistory_.empty() && filterStateHistory_.back()->lastMeasurementTime_ > time)
     {
+      lastHistoryState = filterStateHistory_.back();
       filterStateHistory_.pop_back();
     }
 
+    // If the state history is not empty at this point, it means that our history was large enough, and we
+    // should revert to the state at the back of the history deque.
+    bool retVal = false;
+    if (!filterStateHistory_.empty())
+    {
+      retVal = true;
+      lastHistoryState = filterStateHistory_.back();
```

Comment:
In both the original code and in this PR, in the case where the you had sufficient history, you are setting the `lastHistoryState` to a value that is still in the `filterStateHistory_` container. From this, I am assuming that (a) you will rebuild the `filterStateHistory_` entries from the `measurementQueue_` in some other function, and that (b) the current filter state will not be re-added to the `filterStateHistory_` since it still exists in that container.

However, in the case where there is not sufficient history, the value of `lastHistoryState` does not exist in the `filterStateHistory_`. Does this mean you end up dropping that first state from the `filterStateHistory_` container during the history rebuild?

Reply:
To back way up, before this state reversion code existed, if we got a measurement that was old, we just issued a warning, skipped the prediction step (since the time delta would have been negative), and then corrected.

The `filterStateHistory` only gets updated when we fuse a measurement. We save the state at the time the measurement was fused.

In the original code for this PR, we only bothered reverting the state if there was sufficient history, so we'd back up both the measurement and state histories (which should be parallel) until we reached a time that was <= the one we requested. We would *not* re-enqueue the measurement that was <= the time in question, though, so the `filterStateHistory` would not get any duplicate entries.

When we have insufficient history, it means that we would like to have reverted to some state, but we can't, so we just revert as far back as we can go, and then pass those measurements through, starting with the measurement that made us have to revert to begin with. When we get to that measurement, which is older than any state or measurement we have, we end up with a situation that is pretty much identical to what I described in the first paragraph.

TL; DR, in the first case (sufficient history) we are reverting to the right point in history and then replacing the history values *after* that time as the measurements are processed. In the second case, we are reverting to the oldest point in history we can (which clears the history), and then fusing all the measurements again, which will rebuild the history from the time at which we reverted the state.

I wish I had a whiteboard.

Reply:
I don't want to belabor this point too long, but...

Based on your explanation, then the timestamps associated with the `measurementHistory_`/`measurementQueue_` and the `filterStateHistory_` are synchronized. I am going to assume the `filterStateHistory_` represents the filter state *after* the measurement has been applied.

So, if I receive measurements at t=1 and t=2. `measurementHistory_` contains two entries (for t=1, and t=2), and `filterStateHistory_` also contains 2 entries (t=1, and t=2). If I call `revertTo(1.5s)`, then I reset the state to the value before t=1.5, i.e. the cached state at t=1, and I queue up all measurements received after t=1.5, i.e. just the measurement at t=2. You do not queue up the measurement received at t=1.

However, if I call `revertTo(0.5s)`, then I set the state to the oldest state in `filterStateHistory_`, i.e. the state at time t=1, and a queue up all measurements after t=0.5s, i.e. the measurement at t=1 and the measurement at t=2. Assuming that all of the measurements in the queue will get "replayed", the measurement at t=1 will get applied twice, once when the state was generated and a second time when the measurement was replayed.

Granted, I have not looked at the measurement replay code. So there may be logic that handles such a situation.

Reply:
So the short answer is that yes, the `t1` measurement will get replayed over the state for which it itself was responsible for generating.

However, if we're calling `revertTo(0.5)`, it means we have a measurement `m_old` that arrived with a time stamp of `0.5`. So when we get back to processing measurements, the measurement queue will have three measurements: `m_old`, and the "replayed" measurements from `t1` and `t2`.  The EKF will do an immediate correction on the oldest state using `m_old`, *then* replay the measurement at `t1`, then the measurement at `t2`. So I guess we'd end up with a state that was basically prior state, followed by measurement `t1`, then `m_old`, then measurement `t1` again, then measurement `t2`.

Is there a clean solution to this? Even if I were to revert such that we *didn't* replay the `t1` measurement (and didn't pop the oldest state), we'd still have a state that ended up being a result of fusing `t1`, then `m_old`, then `t2`. `m_old` would still be out of order. I could go either way, really. I'm also not really that concerned, since (a) this would only happen, if at all, once at the beginning of execution (depending on the history length), and (b) it's a degenerate case that I feel users should be responsible for handling.

What do you think?

Reply:
> What do you think?
It's weird. Given that reverting to a time before the beginning of the window should not happen, the details probably don't matter too much.

In my example, you are applying a measurement that occurred at time t=0.5s to the state estimate from time t=1.0s. I don't think anything you do could make that right. I do believe you are "double counting" that one measurement. In general measurements that happen at the same time can be applied to the state estimate in any order. So, in the limit, if we revert to t=0.99999999s, then we would have already applied the measurement at t=1.0. If we then apply just the measurement at t=0.9999999999s we should get the same answer as if we applied the measurements in the opposite, correct order. Apply the measurement at t=1.0s, then the measurement at t=0.999999999s, then the measurement at t=1.0s should give a different answer. For whatever that is worth.

Feel free to merge whenever you are tired of this conversation. :)

Reply:
Also, what was the motivation for including these measurements that are outside of the history window? Instead of the old approach that just ignored them? I mean, I could give you a measurement that was an hour old. If you include that, you will completely mangle the state.

Reply:
What I find is that if I ignore them, then I get ROS Answers questions like "OMG why is my measurement for sensor XYZ being ignored?" It's usually because they have a low-latency sensor mixed with a high-latency one. I could define a threshold, but then it feels kinda' arbitrary (not unlike the way I am handling it now).

Yes, I fully agree that I am double-counting that measurement, but I'm of the mind that this is the corneriest of corner cases for any reasonable data set, and one that I'm comfortable with, unless you feel strongly about it.